scenarios/throughput_stress.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/temporalio/omes/loadgen/throughputstress"
+	"go.temporal.io/sdk/client"
+)
+
+// --option flags
+const (
+	IterFlag     = "internal-iterations"
+	CANEventFlag = "continue-as-new-after-event-count"
+)
+
+const ThroughputStressScenarioIdSearchAttribute = "ThroughputStressScenarioId"
+
+type ThroughputStressStorage struct {
+	workflowCount atomic.Uint64
+}
+
+func init() {
+	loadgen.MustRegisterScenario(loadgen.Scenario{
+		Description: fmt.Sprintf(
+			"Throughput stress scenario. Use --%s and --%s to control internal parameters",
+			IterFlag, CANEventFlag),
+		Executor: &loadgen.GenericExecutor[ThroughputStressStorage]{
+			DefaultConfiguration: loadgen.RunConfiguration{
+				Iterations:    20,
+				MaxConcurrent: 5,
+			},
+			PreScenarioImpl: func(ctx context.Context, info loadgen.ScenarioInfo, storage *ThroughputStressStorage) error {
+				storage = &ThroughputStressStorage{
+					workflowCount: atomic.Uint64{},
+				}
+				attribMap := map[string]enums.IndexedValueType{
+					ThroughputStressScenarioIdSearchAttribute: enums.INDEXED_VALUE_TYPE_KEYWORD,
+				}
+				_, err := info.Client.OperatorService().AddSearchAttributes(ctx,
+					&operatorservice.AddSearchAttributesRequest{
+						Namespace:        info.Namespace,
+						SearchAttributes: attribMap,
+					})
+				var svcErr *serviceerror.AlreadyExists
+				if errors.As(err, &svcErr) {
+					return nil
+				}
+				return err
+			},
+			Execute: func(ctx context.Context, run *loadgen.Run, storage *ThroughputStressStorage) error {
+				internalIterations := run.ScenarioInfo.ScenarioOptionInt(IterFlag, 5)
+				continueAsNewCount := run.ScenarioInfo.ScenarioOptionInt(CANEventFlag, 100)
+				timeout := time.Duration(1*internalIterations) * time.Minute
+
+				wfID := fmt.Sprintf("throughputStress-%s-%d", run.ID, run.IterationInTest)
+				var result throughputstress.WorkflowOutput
+				err := run.ExecuteAnyWorkflow(ctx,
+					client.StartWorkflowOptions{
+						ID:                                       wfID,
+						TaskQueue:                                run.TaskQueue(),
+						WorkflowExecutionTimeout:                 timeout,
+						WorkflowExecutionErrorWhenAlreadyStarted: true,
+						SearchAttributes: map[string]interface{}{
+							ThroughputStressScenarioIdSearchAttribute: run.ScenarioInfo.UniqueRunID(),
+						},
+					},
+					"throughputStress",
+					&result,
+					throughputstress.WorkflowParams{
+						Iterations:                   internalIterations,
+						ContinueAsNewAfterEventCount: continueAsNewCount,
+					})
+				// The 1 is for the final workflow run
+				storage.workflowCount.Add(uint64(result.TimesContinued + result.ChildrenSpawned + 1))
+				return err
+			},
+			PostScenarioImpl: func(ctx context.Context, info loadgen.ScenarioInfo, storage *ThroughputStressStorage) error {
+				totalWorkflowCount := storage.workflowCount.Load()
+				info.Logger.Info("Total workflows executed: ", totalWorkflowCount)
+				return loadgen.VisibilityCountIsEventually(
```

Comment:
If `ExecuteAnyWorkflow` waits for workflow completed anyways, what is the value of this assertion? You're just confirming that you ran as many workflows as you said you did? I could understand if `ExecuteAnyWorkflow` just _started_ workflows and this waited for them all to complete, but otherwise I am not seeing what could fail here unless you're intentionally just integration testing our count feature.

Reply:
It is making sure that visibility counts match what is expected. So yeah, it is testing that. It matters and has found issues before.

Reply:
It's really not a benchmarking tool is the thing. It's a loadgen tool, and it wants to verify some correctness properties about the results of that load. You can definitely argue that maybe that's not the point of a "pure" loadgen tool and I wouldn't really disagree, but, I think the practical reality of what we need is something that will supersede bench-go and ensuring stuff was correct as some basic level is indeed an important part of that.

Reply:
It can be hard to know what to assert against I guess. I guess here this scenario is an e2e test for the visibility count API, but many scenarios may only assert their more primary goals. Technically harmless to assert anything of course. You could test that history is right, that metrics are accurate, etc. I don't mind it here, just seems like an unnecessary step that should never fail if the server tests work properly in the server repo.

Reply:
Without looking at the specific assertion here, just want to say that server teams would definitely get value out of the loadgen failing on invariant violations. I wouldn't ask the loadgen tool to exhaustively check invariants, but I think "if the server tests work properly" is not a safe assumption in reality...